### PR TITLE
chore: Remove enrollmentService.getEnrollment [DHIS2-17712-2]

### PR DESCRIPTION
**Pull request by muilpp**
### Big picture

Tracker has multiple services for each entity like tracked entity, enrollment and event. One is in dhis-api and one in dhis-service-tracker. Our goal is to provide one API (service) per entity that is part of the tracker domain/team.

Trackers architecture splits read/write into exporter services and an importer. So if you want to get data you use an exporter. If you want to insert, update or delete you have to go through the importer. Only then can we ensure integrity of tracker data.

Another goal is that code that provides tracker functionality and is thus owned by the tracker team lives in ideally one maven module (We can settle on a name later on maybe dhis-service-tracker or dhis-tracker).

This is a big task that we are going to implement in many small steps.

### This PR

Moves the method getEnrollment(uid) from the api module to the service tracker module.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/enrollment/DefaultEnrollmentService.java [hunk 1]
```diff
@@ -71,6 +70,11 @@ class DefaultEnrollmentService
 
   private final EnrollmentOperationParamsMapper paramsMapper;
 
+  @Override
+  public Enrollment getEnrollment(String uid) throws ForbiddenException, NotFoundException {
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/enrollment/DefaultEnrollmentService.java:74]: Can you please use our UID type like in the EventService. If its easier you can do it in a separate PR.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java [hunk 2]
```diff
@@ -75,7 +75,10 @@ public TrackerTypeReport deleteEnrollments(TrackerBundle bundle) {
 
       Entity objectReport = new Entity(TrackerType.ENROLLMENT, uid);
 
-      Enrollment enrollment = enrollmentService.getEnrollment(uid);
+      Enrollment enrollment = manager.get(Enrollment.class, uid);
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java:78]: I know I did not for events but I am wondering if we should use the exporter service here @enricocolasante ?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/sms/EnrollmentSMSListener.java [hunk 3]
```diff
@@ -185,23 +190,27 @@ protected SmsResponse postProcess(IncomingSms sms, SmsSubmission submission, Use
     // TODO: Unsure about this handling for enrollments, this needs to be
     // checked closely
     Enrollment enrollment;
-    boolean enrollmentExists = enrollmentService.enrollmentExists(enrollmentid.getUid());
+    boolean enrollmentExists = apiEnrollmentService.enrollmentExists(enrollmentid.getUid());
     if (enrollmentExists) {
-      enrollment = enrollmentService.getEnrollment(enrollmentid.getUid());
+      try {
+        enrollment = enrollmentService.getEnrollment(enrollmentid.getUid());
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/sms/EnrollmentSMSListener.java:196]: You need to pass the user in. This does not have the user set in the context like in an HTTP request context.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/sms/RelationshipSMSListener.java [hunk 4]
```diff
@@ -175,10 +175,13 @@ private RelationshipItem createRelationshipItem(
         break;
 
       case PROGRAM_INSTANCE:
-        Enrollment progInst = enrollmentService.getEnrollment(objId.getUid());
-        if (progInst == null) {
+        Enrollment progInst;
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/sms/RelationshipSMSListener.java:178]:
```suggestion
        Enrollment enrollment;
```

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/EnrollmentServiceTest.java [hunk 5]
```diff
@@ -293,13 +300,14 @@ void shouldNoteDeleteNoteWhenDeletingEnrollment() {
...
-    enrollmentService.addEnrollment(enrollmentA);
+    apiEnrollmentService.addEnrollment(enrollmentA);
 
     assertNotNull(enrollmentService.getEnrollment(enrollmentA.getUid()));
 
-    enrollmentService.deleteEnrollment(enrollmentA);
+    apiEnrollmentService.deleteEnrollment(enrollmentA);
 
-    assertNull(enrollmentService.getEnrollment(enrollmentA.getUid()));
+    Assertions.assertThrows(
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/EnrollmentServiceTest.java:309]: use a static import

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/enrollment/DefaultEnrollmentService.java (hunk 1 above)
Comment (muilpp) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/enrollment/DefaultEnrollmentService.java:74]: I'll do it in the following PR

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java (hunk 2 above)
Comment (muilpp) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java:78]: I think we should use the exporter here, the thing is using the service-tracker `enrollmentService`, we run the validations and map the Enrollment to a new object, which is then used by this class to delete the enrollment, and fails because the session already contains a different Enrollment object with the same id.
Long story short, we need to refactor this logic and it might need its own PR.

Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java:78]: thank you for checking 😄

**Review by teleivo — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java (hunk 2 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/DefaultTrackerObjectsDeletionService.java:78]: I think we need to discuss on how to proceed here.
I am not sure we should use the exporter service here. `deleteEnrollments` method should never be exposed and it should only be used by our importer that already did all the needed validations to check if the enrollment is there and if it can be deleted by the user (and also if the events can be deleted).

Here in the end we should just update the enrollment to `deleted=false` and all the lastUpdated info.

**Review by enricocolasante — APPROVED** [review]
Just a comment on the next steps when we delete
